### ISSUE TITLE
ClientChunkCacheChunkLoadMixin fails to apply when Bobby is installed

### ISSUE THREAD

**Issue by Johni0702** [Communi-Suggestu/Scena#1]
See https://github.com/Johni0702/bobby/issues/157.
Not entirely sure why. Neither mod does anything strange. Bobby literally just injects [once at `HEAD` and once at `RETURN`](https://github.com/Johni0702/bobby/blob/v4.0.1/src/main/java/de/johni0702/minecraft/bobby/mixin/ClientChunkManagerMixin.java#L79-L100).
Looks to just be a bug in the Mixin local capture code. The only reason I'm even forwarding this is because it's trivial to work around on Scena's end because Scena has no reason to use the generally fragile local capture mechanism in the first place, just use `cir.getReturnValue()` to get the chunk.

```
[13:50:15] [Render thread/ERROR]: Mixin apply for mod scena failed scena.mixins.json:platform.client.ClientChunkCacheChunkLoadMixin from mod scena -> net.minecraft.class_631: org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException Unexpecteded NullPointerException whilst transforming the mixin class: [INJECT Applicator Phase -> scena.mixins.json:platform.client.ClientChunkCacheChunkLoadMixin from mod scena -> Apply Injections -> PreInject -> scena.mixins.json:platform.client.ClientChunkCacheChunkLoadMixin from mod scena->@Inject::onChunkReplacedWithPacketData(IILnet/minecraft/class_2540;Lnet/minecraft/class_2487;Ljava/util/function/Consumer;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;ILnet/minecraft/class_2818;Lnet/minecraft/class_1923;)V]
org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException: Unexpecteded NullPointerException whilst transforming the mixin class: [INJECT Applicator Phase -> scena.mixins.json:platform.client.ClientChunkCacheChunkLoadMixin from mod scena -> Apply Injections -> PreInject -> scena.mixins.json:platform.client.ClientChunkCacheChunkLoadMixin from mod scena->@Inject::onChunkReplacedWithPacketData(IILnet/minecraft/class_2540;Lnet/minecraft/class_2487;Ljava/util/function/Consumer;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;ILnet/minecraft/class_2818;Lnet/minecraft/class_1923;)V]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1394) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:1062) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:402) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:327) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422) ~[fabric-loader-0.14.19.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.14.19.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.14.19.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:112) ~[fabric-loader-0.14.19.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at net.minecraft.class_634.method_11120(class_634.java:385) ~[client-intermediary.jar:?]
	at net.minecraft.class_2678.method_11567(class_2678.java:86) ~[client-intermediary.jar:?]
	at net.minecraft.class_2678.method_11054(class_2678.java:20) ~[client-intermediary.jar:?]
	at net.minecraft.class_2600.method_11072(class_2600.java:22) ~[client-intermediary.jar:?]
	at net.minecraft.class_1255.method_18859(class_1255.java:157) [client-intermediary.jar:?]
	at net.minecraft.class_4093.method_18859(class_4093.java:23) [client-intermediary.jar:?]
	at net.minecraft.class_1255.method_16075(class_1255.java:131) [client-intermediary.jar:?]
	at net.minecraft.class_1255.method_5383(class_1255.java:116) [client-intermediary.jar:?]
	at net.minecraft.class_310.method_1523(class_310.java:1129) [client-intermediary.jar:?]
	at net.minecraft.class_310.method_1514(class_310.java:768) [client-intermediary.jar:?]
	at net.minecraft.client.main.Main.method_44604(Main.java:244) [client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:51) [client-intermediary.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:462) [fabric-loader-0.14.19.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.14.19.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) [fabric-loader-0.14.19.jar:?]
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:88) [NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:126) [NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:71) [NewLaunch.jar:?]
Caused by: java.lang.NullPointerException: Cannot read field "desc" because "frame[varInsn.var]" is null
	at org.spongepowered.asm.util.Locals.getLocalsAt(Locals.java:546) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.util.Locals.getLocalsAt(Locals.java:336) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.injection.callback.CallbackInjector.preInject(CallbackInjector.java:479) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.injection.code.Injector.preInject(Injector.java:258) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.preInject(InjectionInfo.java:436) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1371) ~[sponge-mixin-0.12.4+mixin.0.8.5.jar:0.12.4+mixin.0.8.5]
	... 31 more
```

**Comment by marchermans** [comment]
It fails because of locals capture.

**Comment by marchermans** [comment]
I fire the `CHUNK_LOAD` event when the chunk data is updated on the client. This was discussed with fabric and is considered a bug.

I need locals here to properly fire the correct event; as such, any other mixin which is cancellable breaks the local vars table (since that injects a field) and causes the NPE since no var table is available.

I am all ears if you know of a better way to fix this. Due note: this is not a HEAD inject but it injects after the data in memory has been updated from the compound tag, and this behaviour has to stay.

**Comment by Johni0702** [comment]
I don't think I quite follow.

> I need locals here to properly fire the correct event

Why? There may be a subtle difference between local capture and `getReturnValue` that I'm not currently aware of, and if there is, please enlighten me, but afaict based on just that one mixin, you don't.
I suppose it could make a difference if another mod goes and just before your inject replaces the value in the local variable table but not the return value, but that scenario seems highly contrived and I can't think of any reason to do that.

> Due note: this is not a HEAD inject but it injects after the data in memory has been updated from the compound tag, and this behaviour has to stay

I am well aware that this is not a HEAD inject, if it was I wouldn't have suggested using `cir.getReturnValue` because there is no return value at HEAD (there isn't even a Chunk at HEAD), my suggestion can only work at RETURN and it does not involve changing the target.
Just to make it absolutely clear, this is what I am suggesting:
```java
@Mixin(ClientChunkCache.class)
public abstract class ClientChunkCacheChunkLoadMixin extends ChunkSource {

    @Inject(method = "replaceWithPacketData(IILnet/minecraft/network/FriendlyByteBuf;Lnet/minecraft/nbt/CompoundTag;Ljava/util/function/Consumer;)Lnet/minecraft/world/level/chunk/LevelChunk;", at = @At(value = "RETURN", ordinal = 1))
    public void onChunkReplacedWithPacketData(int x, int z, FriendlyByteBuf buffer, CompoundTag tag, Consumer<ClientboundLevelChunkPacketData.BlockEntityTagOutput> consumer, CallbackInfoReturnable<LevelChunk> cir) {
        FabricGameEvents.CHUNK_LOAD.invoker().handle(Minecraft.getInstance().level, cir.getReturnValue());
    }
}
```

> I fire the CHUNK_LOAD event when the chunk data is updated on the client. This was discussed with fabric and is considered a bug.

I feel like there's a typo in here that changes the whole meaning.
Why would that be considered a bug? If it's considered a bug, then why do it?
And how is this related to the workaround and/or issue at hand?

**Comment by poly2it** [comment]
I am also experiencing this issue.

**Comment by Johni0702** [comment]
Should be fixed in Fabric Loader 0.14.20
